contracts/modules/TransferManager/VolumeRestrictionTM.sol

```diff
@@ -0,0 +1,934 @@
+        uint256[] memory passedTimestamps = new uint256[](diffDays);
+        // Check with the bucket and parse all the new timestamps to calculate the sumOfLastPeriod
+        // re-using the local variables to avoid the stack too deep error.
+        (sumOfLastPeriod, fromTimestamp, passedTimestamps) = _bucketCheck(
+            fromTimestamp,
+            diffDays,
+            _from,
+            bucketToUser[_from],
+            false,
+            individualRestriction[_from]
+        );
+        if (_checkValidAmountToTransact(sumOfLastPeriod, _amount, individualRestriction[_from])) {
+            if (_isTransfer) {
+                _updateIndividualStorage(
+                    passedTimestamps,
+                    _from,
+                    fromTimestamp,
+                    _amount,
+                    diffDays,
+                    sumOfLastPeriod
+                );
+            }
+            return Result.NA;
+        } 
+        return Result.INVALID;  
+    }
+
+
+    function _globalTxCheck(
+        uint256 _sumOfLastPeriod,
+        uint256 _dailyAmount,
+        uint256 _amount
+    ) 
+        internal
+        view
+        returns(bool)
+    {
+        // Checking whether the global day restriction is added or not if yes then calculate
+        // the total amount get traded on a particular day (~ _fromTime)
+        if (!_checkValidAmountToTransact(_dailyAmount, _amount, dailyGlobalRestriction) &&
+                now <= dailyGlobalRestriction.endTime && now >= dailyGlobalRestriction.startTime
+        ) {
+            return false;
+        }
+        return _checkValidAmountToTransact(_sumOfLastPeriod, _amount, globalRestriction);
+    }
+
+    function _checkValidAmountToTransact(
+        uint256 _sumOfLastPeriod,
+        uint256 _amountToTransact,
+        VolumeRestriction _restriction
+    ) 
+        internal
+        view
+        returns (bool)
+    {    
+        uint256 _allowedAmount = 0;
+        if (_restriction.typeOfRestriction == RestrictionType.Variable) {
+            _allowedAmount = (_restriction.allowedPercentageOfTokens.mul(ISecurityToken(securityToken).totalSupply()))/ 10 ** 18;
+        } else {
+            _allowedAmount = _restriction.allowedTokens;
+        }
+        // Validation on the amount to transact
+        if (_allowedAmount >= _sumOfLastPeriod.add(_amountToTransact)) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    function _updateIndividualStorage(
+        uint256[] passedTimestamps,
+        address _from,
+        uint256 _fromTime,
+        uint256 _amount,
+        uint256 _diffDays,
+        uint256 _sumOfLastPeriod
+    )
+        internal 
+    {
+        if (_diffDays != 0) {
+            for (uint256 i = 0; i < passedTimestamps.length; i++) {
+                // Add the timestamp that is already passed
+                bucketToUser[_from].timestamps.push(passedTimestamps[i]);
```

Comment:
This will limit the number of days a person can go without transfers to ~180 days. Every push will take ~40k gas. We can reduce the gas usage to ~25k/day by converting the timestamp to uint64 which will allow transfers to happen after at max ~300 days but even then, this is not acceptable in my opinion.